src/linker/Linker.Steps/LinkAttributesParser.cs

```diff
@@ -38,9 +38,10 @@ public void Parse (AttributeInfo xmlInfo)
 
 		static bool IsRemoveAttributeInstances (string attributeName) => attributeName == "RemoveAttributeInstances" || attributeName == "RemoveAttributeInstancesAttribute";
 
-		CustomAttribute[]? ProcessAttributes (XPathNavigator nav, ICustomAttributeProvider provider)
+		(CustomAttribute[]? customAttributes, MessageOrigin[]? origins) ProcessAttributes (XPathNavigator nav, ICustomAttributeProvider provider)
```

Comment:
I would prefer to do this the other way round:
```C#
(CustomAttribute, MessageOrigin)[]? ProcessAttributes(...)
```

Return a single array where each item is the attribute and its origin.

Reply:
This would also simplify the code which uses this.

Reply:
Currently [AddCustomAttributes](https://github.com/dotnet/linker/blob/main/src/linker/Linker/AttributeInfo.cs#L35) accepts `CustomAttribute[]` and places it directly into `CustomAttributes` store. Should we return a single array, we would have to either split it into two to get `CustomAttribute[]` or store `(CustomAttribute, MessageOrigin)[]` instead of `CustomAttribute[]` .